cmake/UnitTesting.cmake

```diff
@@ -19,11 +19,28 @@ include(CTest)
 include(Catch)
 enable_testing()
 
-function(add_blockfactory_test name unit_test_files)
+function(add_blockfactory_test)
+    set(options)
+    set(oneValueArgs NAME)
+    set(multiValueArgs SOURCES)
+
+    set(prefix "abf")
+
+    cmake_parse_arguments(${prefix}
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN})
+
+    set(name ${${prefix}_NAME})
+    set(unit_test_files ${${prefix}_SOURCES})
+
     set(targetname ${name}UnitTests)
     add_executable(${targetname}
-        ${unit_test_files})
+        "${unit_test_files}")
 
+    target_compile_definitions(${targetname} PRIVATE
+        CMAKE_BINARY_DIR="${CMAKE_BINARY_DIR}")
```

Comment:
What is the rationale for this? Note that `CMAKE_BINARY_DIR` is one of the variables that may have an unexpected values if BlockFactory is downloaded and compiled using the FetchContent module, and added to the build via an `add_subdirectory` (see https://github.com/robotology/ycm/issues/162).

Reply:
This is helpful for the tests. I have a staging PR with a test that has to load a plugin from the build directory. If I have this variable defined, there's no need to alter the environment since the plugin path can be added directly from the code. Switching to `PROJECT_BINARY_DIR` would suffice?

Reply:
Mmh not sure if that would work in the use case I had in mind. Have a look at:

https://github.com/diegoferigo/blockfactory/blob/beed55a1a1b86fedfb90ecb0deffd0ec8639c8a0/tests/Factory/FactoryUnitTest.cpp#L49-L51

In this case I know that the plugin library will be stored in `${PROJECT_BINARY_DIR}/lib`. Using the generator `$<TARGET_FILE_DIR:tgt>` would not work because the plugin could be created in another target (that btw being a plugin is not even linked to the unit test target) and its destination is hard to predict if not from knowing how the CMake project was configured.

I think that the ignition libraries do something similar to this approach, refer to [SystemLoader_TEST.cc](https://bitbucket.org/ignitionrobotics/ign-gazebo/src/90cf23937f6add02da3a710d16eda700d691a3ab/src/SystemLoader_TEST.cc#lines-37).

Reply:
> In this case I know that the plugin library will be stored in ${PROJECT_BINARY_DIR}/lib. 

Nitpick: in this case you know that on *nix system and with a single-config CMake generator, the plugin library will be stored in `${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR}"` (see https://github.com/robotology/blockfactory/blob/ad27cec5eb1c6ac633f549f26b10e40c8faf2167/CMakeLists.txt#L100).  Regarding the use of `BlockFactory` inside another project via `add_subdirectory`, then in this case using `CMAKE_BINARY_DIR` is correct, as that will be the plugin location even when the `BlockFactory` library will be embedded in another project. 

> I think that the ignition libraries do something similar to this approach, refer to [SystemLoader_TEST.cc](https://bitbucket.org/ignitionrobotics/ign-gazebo/src/90cf23937f6add02da3a710d16eda700d691a3ab/src/SystemLoader_TEST.cc#lines-37).

That approach will fail in the following cases: 
* Plugin that are not `LIBRARY`, but `RUNTIME` (that are installed in `/bin`), i.e. Windows
* Multiple-config generators, for which the library is installed in `${Config}` subdirectory (typically `Release` or `Debug`) i..w Visual Studio or XCode 
* As soon as the user has a `CMAKE_INSTALL_LIBDIR` different from `lib` (see https://github.com/Kitware/CMake/blob/v3.13.3/Modules/GNUInstallDirs.cmake#L203), i.e. in Debian systems when installing in `/usr` 

> Using the generator `$<TARGET_FILE_DIR:tgt>` would not work because the plugin could be created in another target (that btw being a plugin is not even linked to the unit test target) and its destination is hard to predict if not from knowing how the CMake project was configured.

Sorry, I think I was not super-clear. I meant use the `TARGET_FILE_DIR` generator expression to pass the location of the plugin target, see `$<TARGET_FILE_DIR:plugin_library_target>`.